_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+
+In this application, the user hits a **Register** button and the server responds with information about the origin of the registration (hostname and origin), a byte-based user handle for identification, a randomly generated challenge nonce, and other optional information about the type of registration the server will accept.
+
+The browser receives this information. A JavaScript function on the client re-encodes and formats the data for the WebAuthn API, and the generated JavaScript object is passed to the browser's `navigator.credentials.create()` function.
+
+The WebAuthn API passes the request to a connected authenticator, which normally requests some form of user verification (such as a PIN number or fingerprint), then generates a key pair for future verification and a credential ID to scope the relying party. The public key, credential ID, and challenge nonce, along with optional additional information, are passed back to the service provider; the provider checks the challenge nonce for tampering and then stores the public key and credential ID for future use.
+
+{% img blog/webauthn-java/auth-flow-diagram.png alt:"WebAuthn authentication flow" width:"800" %}{: .center-image }
+
+Authentication functions similarly, though the content of the data passed around is slightly different. The user hits a **Sign-in** button and the service provider sends identifying information (credential ID and a challenge nonce). JavaScript on the client passes that request to the authenticator using the WebAuthn API, then the authenticator verifies the credential and returns a key, which is passed back to the service provider. The service provider verifies the key and performs any steps necessary to grant the user access to appropriate resources, like setting authentication cookies.
+
+## How to authenticate using WebAuthn in Java
+
+In this article, you're going to build a Java application that authenticates users using WebAuthn. To bootstrap a Java web framework that can create and store user sessions, you'll use [Spring Boot](https://spring.io/projects/spring-boot) to implement the web application, then implement the WebAuthn server, built by the hardware authentication manufacturer [Yubico](https://www.yubico.com/), to create the necessary JSON for the client to create and use credentials.
+
+The final version of the code can be found [here](https://github.com/oktadev/webauthn-java-example). The demo application uses Java 17 and Apache Maven for dependency management.
+
+## Getting started
+
+First, go to the [Spring Initializr page](https://start.spring.io/#!type=maven-project&language=java&&packaging=jar&jvmVersion=17&groupId=com.webauthn&artifactId=app&name=app&description=Demo%20Java%20project%20for%20WebAuthN%20Authentication&packageName=com.webauthn.app&dependencies=lombok,data-jpa,h2,web,thymeleaf) and add your dependencies:
+
+* **Lombok**: A library for automatically generating Initializer classes, getters, and setters for speed of development.
```

Comment:
I don't know if "Initializer" should be here. It might cause confusion because "Spring Initializr".

Reply:
Good point!
```suggestion
* **Lombok**: A library used to speed up development by automatically generating constructors, getters and setters.
```